### PR TITLE
[AUR-185] Added event communication models

### PR DESCRIPTION
**Pull request by Ficzer**
Note: All models are the same, one model covers all requirements

### CODE REVIEW

File: requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+void eventbus_removeEmptySubscription(int emptySubscriptionIndex)
+{
+    for (int j = emptySubscriptionIndex; j < subscriptionCounter || j < SUBSCRIPTION_TABLE_BUFFER_SIZE - 1; j++)
+    {
+         subscriptionTable[j] = subscriptionTable[j + 1];
+    }
+}
+
+void eventbus_updateRoutingTable(asn1SccT_UInt32 eventId)
```
Comment (Lurkerpas) [inline requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c]: enable_routes_for_event?

File: requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventlistener2/C/src/eventlistener2.c [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    nanoseconds += (uint32_t)IN_eventmessage->timestamp.f1 << 24;
+    nanoseconds += (uint32_t)IN_eventmessage->timestamp.f2 << 16;
+    nanoseconds += (uint32_t)IN_eventmessage->timestamp.f3 << 8;
+
+    time = SEC_TO_NS((uint64_t)seconds) + (uint64_t)nanoseconds;
+
+    printf("Event received by Listener 2:\n");
+    printf("  event id =       %ld\n", IN_eventmessage->id);
+    switch(IN_eventmessage->eventType)
```
Comment (Lurkerpas) [inline requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventlistener2/C/src/eventlistener2.c]: Normally a bit of code duplication in test models is not a problem, but have you considered providing a lib/header for handling the timestamps and event printing?

File: requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+void eventbus_PI_receive_event
+      (const asn1SccT_EventMessage *IN_eventmessage)
+
+{
+    eventbus_updateRoutingTable(IN_eventmessage->id);
+
+    eventbus_RI_dispatch_event(IN_eventmessage);
+
+    eventbus_dispatch_event_set_all_routes_enabled(false);
```
Comment (Lurkerpas) [inline requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c]: Not exactly the fastest solution. In the future please consider using e.g., 2D array. While the current solution is more portable, use of C++ could be considered (hashmaps for better memory utilization). updateRoutingTable could be modified to turn the routes both on and off, depending on the event. Resetting all to enabled wouldn't be necessary then.

File: requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c (hunk 1 above)
Comment (Ficzer) [reply requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c]: Done

File: requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventlistener2/C/src/eventlistener2.c (hunk 2 above)
Comment (Ficzer) [reply requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventlistener2/C/src/eventlistener2.c]: Done, added libs

File: requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c (hunk 3 above)
Comment (Ficzer) [reply requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c]: Done

File: requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    asn1SccT_UInt32 eventId;
+} Subscription;
+
+static int subscriptionCounter;
+static Subscription subscriptionTable[SUBSCRIPTION_TABLE_BUFFER_SIZE] = {{0}};
+
+
+int eventbus_getSubscriptionIndex(const asn1SccPID pid, const asn1SccT_UInt32 eventId)
+{
+    for(int i = 0; i < subscriptionCounter || i < SUBSCRIPTION_TABLE_BUFFER_SIZE; i++)
```
Comment (Lurkerpas) [inline requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c]: Shouldn't this be &&?

File: requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        {
+            return i;
+        }
+    }
+    return -1; // Subscription is not present
+}
+
+void eventbus_removeEmptySubscription(int emptySubscriptionIndex)
+{
+    for (int j = emptySubscriptionIndex; j < subscriptionCounter || j < SUBSCRIPTION_TABLE_BUFFER_SIZE - 1; j++)
```
Comment (Lurkerpas) [inline requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c]: Shouldn't this be &&? Also, maybe we could just ensure that subscriptionCounter  is <= than SUBSCRIPTION_TABLE_BUFFER_SIZE (also for the previous comment)

Comment (Lurkerpas) [reply requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c]: assert would probably do just fine....

File: requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c (hunk 4 above)
Comment (Ficzer) [reply requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c]: Removed check

File: requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c (hunk 5 above)
Comment (Ficzer) [reply requirements/EVENT_SERVICES_AURORA_0001/register-event-listeners/work/eventbus/C/src/eventbus.c]: Constraint is present in line 94

> if(subscriptionCounter == SUBSCRIPTION_TABLE_BUFFER_SIZE - 1)
   {
        return;
   }

**Review by Lurkerpas — APPROVED** [review]

**Pull request merged**